### DISCUSSION TITLE
Four. Isolating the Domain

### DISCUSSION THREAD

**Discussion by Mehdi-H** [Mehdi-H/DDD-blue-book-collaborative-reading#5] [Chapter]
# Four. Isolating the Domain

You can discuss here about this chapter ! 👇

Ask questions,
or share anything you find insightful !

**Comment by sofcalca** [comment]
## :construction: Glossary

LAYERED ARCHITECTURES = an architecture that isolates the domain (domain layer) from other parts of the code (user interface, application layer, infrastructure layer)
OBSERVERS =
MODEL-VIEW-CONTROLLER = 
MODEL-VIEW SEPARATION PATTERN = 
APPLICATION COORDINATOR =
SERVICES =

## What I learned in this chapter

- Isolate the domain from other parts of the code via a layered architecture to make it easier to find, read, and reason about it: it is **the isolation of the domain that makes model-driven design possible**
- This approach brings complexity to the software and is not always the right approach to take
    - simple projects do not need this level of layering, because the domain is not that complex and can be understood without isolating it
    - layering the code will bring overhead
    - needs strong skills
    - ROI of layering is not profitable
- Migrating to a different code architecture approach is costly

**Reply by Mehdi-H** [reply]
> simple projects do not need this level of layering, because the domain is not that complex and can be understood without isolating it

- What would you call a simple project ?
- At the begining of a project, are you able to state that is "hard enough" to do DDD ? Or do you need to start a DDD approach to be able then to make this statement ?

I have the feeling that your bullets points describe a simple project _a posteriori_, and I wouldn't know for example measuring the ROI of layering before trying it on a new projet with a model I don't know yet

**Comment by sofcalca** [comment]
This trade-off between designing a software with layered architecture vs without makes me think about [this graph](https://twitter.com/mathiasverraes/status/1174746248059879424/photo/2) comparing the cost of:
- the complexity brought by the implementation of DDD (plus learning it)
- an "easy reductionist solution"
For a given intrinsic business complexity over time.

So I'd say that DDD is over-engineering when:
- the complexity of the domain is low
- it is a short-lived software (a POC) - but it then should be disposable and disposed if the domain is complex and the POC wants to be generalized

**Reply by Mehdi-H** [reply]
Just displaying the graph you mentioned 📈 

![Complexity vs time](https://pbs.twimg.com/media/EE2Je0qU4AAlxjd?format=jpg&name=small)[^1]

[^1]: Credit [Mathias Verraes](https://twitter.com/mathiasverraes) and [Wassel Alazhar](https://twitter.com/wasselovski) (_aka_ [jcraftsman](https://github.com/jcraftsman)) @ ExploreDDD2019, [published on Twitter 🐦 ](https://twitter.com/mathiasverraes/status/1174746248059879424)
